File: apps/intelligence/marketrift_intelligence/analysis_job.py
import json
from pathlib import Path

from jsonschema import Draft202012Validator, FormatChecker

SCHEMA_PATH = Path(__file__).resolve().parents[3] / "packages/contracts/analyze-document-job.v1.schema.json"
VALIDATOR = Draft202012Validator(json.loads(SCHEMA_PATH.read_text(encoding="utf-8")), format_checker=FormatChecker())


def make_analysis_job(tenant_id: str, document_id: str, extractor_version: str) -> dict:
    job = {
        "version": 1,
        "tenant_id": tenant_id,
        "document_id": document_id,
        "extractor_version": extractor_version,
        "idempotency_key": f"analysis-{document_id}-{extractor_version}",
    }
    return validate_analysis_job(job)


def validate_analysis_job(payload: object) -> dict:
    VALIDATOR.validate(payload)
    assert isinstance(payload, dict)
    expected = f"analysis-{payload['document_id']}-{payload['extractor_version']}"
    if payload["idempotency_key"] != expected:
        raise ValueError("analysis idempotency key does not match document and version")
    return payload

File: apps/api/src/analysis-job.ts
import { readFileSync } from 'node:fs';
import { join } from 'node:path';
import Ajv2020 from 'ajv/dist/2020';
import addFormats from 'ajv-formats';

export interface AnalyzeDocumentJobV1 {
  version: 1;
  tenant_id: string;
  document_id: string;
  extractor_version: string;
  idempotency_key: string;
}

export const activeExtractorVersion = process.env.EXTRACTOR_VERSION ?? 'review-issues-v1';
const ajv = new Ajv2020({ strict: false });
addFormats(ajv);
const schema = JSON.parse(readFileSync(join(__dirname, '../../../packages/contracts/analyze-document-job.v1.schema.json'), 'utf8')) as object;
const validate = ajv.compile<AnalyzeDocumentJobV1>(schema);

export function makeAnalysisJob(tenantId: string, documentId: string): AnalyzeDocumentJobV1 {
  const job: AnalyzeDocumentJobV1 = {
    version: 1, tenant_id: tenantId, document_id: documentId,
    extractor_version: activeExtractorVersion,
    idempotency_key: `analysis-${documentId}-${activeExtractorVersion}`,
  };
  if (!validate(job)) throw new Error(`Invalid analysis job contract: ${ajv.errorsText(validate.errors)}`);
  return job;
}
